### PR TITLE
Renaming `DataOptions` to `DataLocale`

### LINKED ISSUE #1995: Rename ResourceOptions to ResourceLocale/DataLocale
I've never been super happy with the name "resource options". It occurred to me that now that we've settled on a locale-like data model, maybe we should rename it to ResourceLocale or maybe DataLocale.

This would make the type hierarchy look like:

```rust
// Public fields
#[allow(clippy::exhaustive_structs)] // this type is stable
pub struct DataRequest {
    pub locale: DataLocale,
    pub metadata: DataRequestMetadata,
}

// Public fields
#[non_exhaustive]
pub struct DataRequestMetadata { ... }

// Private fields
pub struct DataLocale {
    langid: LanguageIdentifier,
    keywords: unicode_ext::Keywords,
}
```

All of the existing conversions would still work: you can convert a Locale to a DataLocale and back.

Needs approval from:

- [x] @Manishearth 
- [ ] @zbraniecki

### LINKED ISSUE #2117: Fold MissingLocale and MissingVariant into MissingResourceOptions
The distinction does not seem valuable, and it makes the fallback provider a bit more complicated.

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #1995 
Fixes #2117

### CODE REVIEW

File: provider/datagen/src/transform/cldr/datetime/mod.rs [hunk 1]
```diff
@@ -55,16 +55,16 @@ macro_rules! impl_data_provider {
...
                     req: &DataRequest,
                 ) -> Result<DataResponse<$marker>, DataError> {
-                    let langid = req.options.get_langid();
+                    let langid = req.locale.get_langid();
                     let calendar = req
-                        .options
+                        .locale
                         .get_unicode_ext(&key!("ca"))
-                        .ok_or_else(|| DataErrorKind::NeedsVariant.into_error())?;
+                        .ok_or_else(|| DataErrorKind::MissingLocale.into_error())?;
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/datetime/mod.rs:62]: Nit: `NeedsVariant` should probably collapse to `NeedsLocale`, not `MissingLocale`

**Review by sffc — APPROVED** [review]
thanks!

File: provider/datagen/src/transform/cldr/datetime/mod.rs (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/datetime/mod.rs:62]: mmh it's unclear. We could have a locale here without the unicode extension, which would just be an unsupported locale

Comment (robertbastian) [reply provider/datagen/src/transform/cldr/datetime/mod.rs:66]: So I'm checking if the locale is empty first, which throws `NeedsLocale`, and then if the extension is missing or we don't support the calendar it throws `MissingLocale`

**Review by sffc — APPROVED** [review]

**Pull request merged**